In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
dataset = [
    ['milk', 'bread', 'nuts', 'apple'],
    ['milk', 'bread', 'nuts'],
    ['milk', 'bread'],
    ['milk', 'apple'],
    ['bread', 'apple'],
    ['milk', 'bread', 'apple'],
    ['milk', 'bread', 'apple'],
    ['bread', 'nuts']
]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
te= TransactionEncoder()
te_array = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_array, columns = te.columns_)

In [ ]:
print("One-Hot Encoded Transaction Data:")
print(df)

One-Hot Encoded Transaction Data:
   apple  bread   milk   nuts
0   True   True   True   True
1  False   True   True   True
2  False   True   True  False
3   True  False   True  False
4   True   True  False  False
5   True   True   True  False
6   True   True   True  False
7  False   True  False   True


In [ ]:
frequent_itemsets = apriori(df, min_support = 0.3, use_colnames = True)
print("\n Frequent Itemsets:")
print(frequent_itemsets)


 Frequent Itemsets:
   support              itemsets
0    0.625               (apple)
1    0.875               (bread)
2    0.750                (milk)
3    0.375                (nuts)
4    0.500        (bread, apple)
5    0.500         (milk, apple)
6    0.625         (milk, bread)
7    0.375         (nuts, bread)
8    0.375  (milk, bread, apple)


In [ ]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold = 1.0)
print("\n Association Rules:")
print(rules[['antecedents','consequents','support','confidence','lift']])


 Association Rules:
      antecedents     consequents  support  confidence      lift
0          (milk)         (apple)    0.500    0.666667  1.066667
1         (apple)          (milk)    0.500    0.800000  1.066667
2          (nuts)         (bread)    0.375    1.000000  1.142857
3         (bread)          (nuts)    0.375    0.428571  1.142857
4  (bread, apple)          (milk)    0.375    0.750000  1.000000
5          (milk)  (bread, apple)    0.375    0.500000  1.000000


In [ ]:
rules = rules.sort_values(by='lift', ascending=False)
print("\n Top Rules by Lift:")
print(rules[['antecedents','consequents','support','confidence','lift']])


 Top Rules by Lift:
      antecedents     consequents  support  confidence      lift
2          (nuts)         (bread)    0.375    1.000000  1.142857
3         (bread)          (nuts)    0.375    0.428571  1.142857
0          (milk)         (apple)    0.500    0.666667  1.066667
1         (apple)          (milk)    0.500    0.800000  1.066667
4  (bread, apple)          (milk)    0.375    0.750000  1.000000
5          (milk)  (bread, apple)    0.375    0.500000  1.000000


### Observations:

1.  **Frequent Itemsets**: The analysis identified frequent itemsets based on a minimum support of 0.3. This means that combinations of items appearing in at least 30% of the transactions are considered significant. For example, 'milk' has a support of 0.75, indicating it appears in 75% of the transactions.

2.  **Association Rules**: The association rules were generated with a minimum lift of 1.0. Lift measures how much more likely an item is to be purchased given that another item is purchased, relative to their individual purchase rates. A lift greater than 1 indicates a positive association.
    *   The rule `(nuts) -> (bread)` has the highest lift (1.14), suggesting that customers who buy nuts are more likely to buy bread, and this relationship is stronger than would be expected by chance.
    *   Rules like `(milk) -> (apple)` and `(apple) -> (milk)` also show a positive association with a lift of 1.06, meaning these items are moderately associated.
    *   Rules with a lift of 1.0, like `(bread, apple) -> (milk)` and `(milk) -> (bread, apple)`, indicate that these items are bought together as often as would be expected by chance, given their individual frequencies.